File: Data Structures & Algorithms/eating-bananas/submission-6.go
import "slices"

func rate(r int, piles []int) int {
    count := 0
    for _, pile := range piles {
        count += (pile + r - 1) /r
    }
    return count
}

func minEatingSpeed(piles []int, h int) int {
    left := 1
    right := slices.Max(piles)
    res := right

    for left <= right {
        mid := left + (right - left) / 2

        if rate(mid, piles) <= h {
            res = mid
            right = mid - 1
        } else {
            left = mid + 1
        }
    }

    return res 	
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-6.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        dp = [0] * 26
        l, res = 0, 0
        maxx =0

        for r in range(len(s)):
            ch = ord(s[r]) - ord("A")
            dp[ch] += 1
            maxx = max(maxx, dp[ch])
            w = r - l + 1

            while w - maxx > k:
                dp[ord(s[l]) - ord("A")] -= 1
                l += 1
                w = r - l + 1

            res = max(res, w)

        return res